### PR TITLE
public alloc, ValidAsZeroBits for tuples, CudaSlice slicing with any range

### LINKED ISSUE #54: alloc (without memsetting to 0) should be public
can this function be public?
https://github.com/coreylowman/cudarc/blob/2785abba641f3bff4234e0aa70542deab6f67c86/src/driver/safe.rs#L388

Sometimes i know that i'll overwrite/initialize the memory, but memsetting 2GB-3GB before is rather a waste of time.

### PR DESCRIPTION
**Pull request by M1ngXU**
`CudaView` should have the same operations as `CudaSlice` (like slicing, copying, ...), though this would be a seperate PR. This closes #54, #55 (though a better name is required), #56 (partially) and #57.

### CODE REVIEW

File: src/driver/safe.rs [hunk 1]
```diff
@@ -273,15 +302,12 @@ impl<T> CudaSlice<T> {
...
-            })
-        } else {
-            None
-        }
+    pub fn try_slice(&self, range: impl RangeBounds<usize>) -> Option<CudaView<'_, T>> {
+        range.bounds(..self.len()).map(|(start, end)| CudaView {
+            ptr: self.cu_device_ptr + (start * std::mem::size_of::<T>()) as u64,
+            slice: self,
+            len: 1 + end - start,
+        })
```
Comment (coreylowman) [inline src/driver/safe.rs:310]: This is super cool, nice helper trait!!

File: src/driver/safe.rs [hunk 2]
```diff
@@ -1230,7 +1314,9 @@ extern \"C\" __global__ void sin_kernel(float *out, const float *inp, size_t num
 
         for i in 0..5 {
             let a_sub = a_dev.try_slice(i * 2..).unwrap();
+            assert_eq!(a_sub.len, 10 - 2 * i);
             let mut b_sub = b_dev.try_slice_mut(i * 2..).unwrap();
+            assert_eq!(b_sub.len, 10 - 2 * i);
```
Comment (coreylowman) [inline src/driver/safe.rs:1319]: Want to adjust this test to use `i*2..i*2+2`?

**Review by coreylowman — APPROVED** [review]

File: src/driver/safe.rs (hunk 2 above)
Comment (M1ngXU) [reply src/driver/safe.rs:1319]: hmm, this is actually an existing test where i just added to test the length (while this should probably be seperated ...). But what exactly do you mean? the slicing with i*2..i*2+2; if yes, why?

Comment (coreylowman) [reply src/driver/safe.rs:1319]: Mainly want something to test that the length of CudaView is properly used. I think this can be done in #56

**Pull request merged**